### DISCUSSION TITLE
How to SPARQL on local or remote RDF data ?

### DISCUSSION THREAD

**Discussion by tfrancart** [comunica/comunica#605] [Q&A] [answered]
#### Issue type:

- :question: Question

____
#### Question:

Hello

I am trying to parse and query with SPARQL either a local RDF string (i.e. in the browser page) or an RDF file located in a remote file.
I would like the parsing to be done once, and then query the triples/quads with SPARQL.
I am confused by https://github.com/comunica/comunica/tree/master/packages/actor-init-sparql-rdfjs#readme, and I don't understand how to hook together the parsing of an RDF string, a Store (N3Store ?), and an Engine querying the store.
Should I use `actor-init-sparql-file` instead of rdfjs ? it seemed to me that this would reparse the same file each time I am making a query, which is what I want to avoid.

Any help appreciated !
Thanks for this valuable work

Thomas

**Comment by rubensworks** [comment, marked as answer]
If you want to re-query the same data several times, then `@comunica/actor-init-sparql-rdfjs` would indeed be the way to go.

So what you'd want to do, is parsing your RDF file(s), and importing it into your `N3Store`.
The easiest way to do this, is by using something like [`rdf-parse`](https://github.com/rubensworks/rdf-parse.js) and [`rdf-store-stream`](https://github.com/rubensworks/rdf-store-stream.js) as follows:

```javascript
import rdfParser from "rdf-parse";
import {storeStream} from "rdf-store-stream";
const newEngine = require('@comunica/actor-init-sparql-rdfjs').newEngine;

// Create an RDF stream
const quadStream = rdfParser.parse(fs.createReadStream('cartoons.ttl'), { contentType: 'text/turtle' });

// Import the stream into a store
const store = await storeStream(quadStream);

// Use the store
const myEngine = newEngine();
const result = await myEngine.query('SELECT * { ?s ?p <http://dbpedia.org/resource/Belgium>. ?s ?p ?o } LIMIT 100',
  { sources: [ { type: 'rdfjsSource', value: store } ] });
result.bindingsStream.on('data', (data) => {
  // Each data object contains a mapping from variables to RDFJS terms.
  console.log(data.get('?s'));
  console.log(data.get('?p'));
  console.log(data.get('?o'));
});
```

**Comment by tfrancart** [comment]
Thanks ! populating the store was the part I was missing.
